Add maps to ExcisionSpheres

knelli2 opened the pull request:
## Proposed changes

Discussion of why we want this in #5150. Fixes #5180.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Move ExcisionSphere from DomainStructure to Domain

Coordinate maps will be added to ExcisionSphere so we don't want
cyclic dependencies

knelli2 committed:
Add grid to inertial maps to ExcisionSphere

wthrowe commented on tests/Unit/Domain/Test_ExcisionSphere.cpp:
```diff
@@ -149,6 +154,35 @@ void test_error() {
+  CHECK(excision_sphere.is_time_dependent());
+  CHECK(excision_sphere.moving_mesh_grid_to_inertial_map() == *time_dep_map);
+
+  const ExcisionSphere<3> copy_assigned_excision_sphere = excision_sphere;
```
That's not a copy assignment.  You need to assign to a preexisting object.

knelli2 replied:
Oops 😅

wthrowe commented on tests/Unit/Domain/Test_ExcisionSphere.cpp:
```diff
@@ -149,6 +154,35 @@ void test_error() {
+
+  ExcisionSphere<3> copied_excision_sphere{excision_sphere};
+  CHECK(copied_excision_sphere == excision_sphere);
+  copied_excision_sphere = excision_sphere;
```
Don't start with the correct answer.  The usual way to screw these up is to forget to assign a field, and this won't catch that.  Just default-construct one and then assign to it.

knelli2 replied:
Third times a charm

nilsvu replied:
Can you just use `test_copy_semantics`?

nilsvu commented on src/Domain/ExcisionSphere.cpp:
```diff
@@ -24,8 +24,9 @@ ExcisionSphere<VolumeDim>::ExcisionSphere(
-  ASSERT(radius_ > 0.0,
-         "The ExcisionSphere must have a radius greater than zero.");
+  if (radius_ <= 0.0) {
+    ERROR("The ExcisionSphere must have a radius greater than zero.");
```
Can you hit this by making bad decisions in an input file? That's when I use `ERROR`. For things that are programming errors use `ASSERT`.

knelli2 replied:
That's dependent on how the domain creator uses this. I don't see an issue with an extra check in the constructor. And this object doesn't make sense with a negative radius so why should we rely on external code to ensure a correct radius is passed in?

nilsvu commented on src/Domain/ExcisionSphere.cpp:
```diff
@@ -29,6 +32,56 @@ ExcisionSphere<VolumeDim>::ExcisionSphere(
+template <size_t VolumeDim>
+ExcisionSphere<VolumeDim>::ExcisionSphere(
+    const ExcisionSphere<VolumeDim>& rhs) {
+  radius_ = rhs.radius_;
+  center_ = rhs.center_;
+  abutting_directions_ = rhs.abutting_directions_;
+  if (rhs.grid_to_inertial_map_ == nullptr) {
+    grid_to_inertial_map_ = nullptr;
+  } else {
+    grid_to_inertial_map_ = rhs.grid_to_inertial_map_->get_clone();
+  }
```
```cpp
  *this = rhs;
```

nilsvu commented on src/Domain/ExcisionSphere.cpp:
```diff
@@ -29,6 +32,56 @@ ExcisionSphere<VolumeDim>::ExcisionSphere(
+  // calling this as often as the coordinate maps in the Blocks, so always
+  // checking shouldn't be costly (and we wrap in an unlikely for good measure).
+  if (UNLIKELY(grid_to_inertial_map_ == nullptr)) {
+    ERROR(
```
same comment about user vs programmer error

knelli2 replied:
I've messed this up when coding things into the domain creator and then an ASSERT wasn't hit in release mode and I got a segfault/UB. If this was going to be called super often like in the RHS, I'd make it an ASSERT, but this will only be called on events so it shouldn't be slow. And for that reason, I'd always want the extra protection.

nilsvu replied:
Just test things in Debug mode, that's what it's for :) And that's why the distinction between ERROR and ASSERT exist.

nilsvu commented on src/Domain/ExcisionSphere.cpp:
```diff
@@ -60,13 +113,29 @@ void ExcisionSphere<VolumeDim>::pup(PUP::er& p) {
-         lhs.abutting_directions() == rhs.abutting_directions();
+  // First check time dep maps. If both are time dep, safe to dereference
+  // unique_ptr
+  if ((lhs.is_time_dependent() and not rhs.is_time_dependent()) or
+      ((not lhs.is_time_dependent()) and rhs.is_time_dependent())) {
```
```cpp
  if (lhs.is_time_dependent() != rhs.is_time_depentend()) {
```

nilsvu commented on src/Domain/ExcisionSphere.hpp:
```diff
@@ -46,17 +53,37 @@ class ExcisionSphere {
+
+  /// \brief The map going from the last time independent frame to the frame in
+  /// which the equations are solved. Only used when the coordinate map is
+  /// time-dependent.
```
"Only use when `is_time_dependent()` is true"

knelli2 committed:
Update error test in ExcisionSphere test

knelli2 committed:
Use ExcisionSpheres in SeparationLessThan

knelli2 commented:
Rebased, squashed in the changes into the first 3 commits (because they were small and there weren't that many), and added a new commit that uses ExcisionSphere maps in the SeparationLessThan trigger

nilsvu commented on src/Evolution/Triggers/SeparationLessThan.cpp:
```diff
@@ -13,52 +13,53 @@
+
+  const auto check_excision_sphere =
+      [&excision_spheres]([[maybe_unused]] const std::string& object) {
+        ASSERT(excision_spheres.count("ExcisionSphere" + object) == 1,
```
So.. this is ironic but these should be `ERROR`s I think 😅 Because they're caused by choices in the input file (choosing a domain that doesn't have two excisions, or that don't have time dependences). Ideally the check would happen at option-construction, but I don't see a great way of doing that ATM.

knelli2 replied:
🤦🏻

wthrowe commented on src/Evolution/Triggers/SeparationLessThan.cpp:
```diff
@@ -13,52 +13,56 @@
+      domain.excision_spheres();
+
+  const auto check_excision_sphere =
+      [&excision_spheres]([[maybe_unused]] const std::string& object) {
```
Drop "maybe_unused".  It's always used.

wthrowe reviewed (changes requested):
Squash, including this.

knelli2 commented:
Squashed and rebased
